File: drishti-mvp-functions/analyzeWithVertexVision.js
const { onObjectFinalized } = require("firebase-functions/v2/storage");
const { logger } = require("firebase-functions");
const { db, admin, geminiModel } = require('./index.js').dependencies;
const { Storage } = require("@google-cloud/storage");
const vision = require("@google-cloud/vision").v1;

const storage = new Storage();
const client = new vision.ImageAnnotatorClient();

module.exports.analyzeWithVertexVision = onObjectFinalized(
  {
    bucket: "project-drishti-central1-bucket-vision",
    memory: "2GiB",
    timeoutSeconds: 300,
  },
  async (event) => {
    const object = event.data;
    const filePath = object.name;
    const bucketName = object.bucket;
    logger.log(`New image uploaded: ${filePath}`);

    try {
      // Lookup media doc to get zoneId
      const mediaSnapshot = await db
        .collection("media")
        .where("objectPath", "==", filePath)
        .get();
      if (mediaSnapshot.empty) {
        logger.log(`⚠️ No Firestore media doc found for ${filePath}`);
        return;
      }
      const mediaDoc = mediaSnapshot.docs[0];
      const mediaData = mediaDoc.data();
      const zoneId = mediaData.zone || "Unknown";
      const publicImageUrl = mediaData.fileUrl || `https://storage.googleapis.com/${bucketName}/${filePath}`;

      // Download the image
      const file = storage.bucket(bucketName).file(filePath);
      const [buffer] = await file.download();

      // Call Vertex AI Vision for object localization
      const [result] = await client.objectLocalization({
        image: { content: buffer },
      });
      const personCount = result.localizedObjectAnnotations.filter(obj => obj.name === 'Person').length;
      logger.log(`Detected ${personCount} people in ${filePath}`);

      // Prepare base64 image for Gemini
      const base64Image = buffer.toString("base64");
      const mimeType = object.contentType || "image/jpeg";

      // Compose Gemini prompt (same as analyzeWithGemini.js, but include personCount)
      const prompt = `Analyze the image for the following and return ONLY this JSON:\n{
  "peopleCount": <number>,
  "crowdDensity": "low|moderate|high",
  "smokeDetected": <true|false>,
  "fireDetected": <true|false>,
  "medicalEmergency": <true|false>,
  "potentialRisk": <true|false>,
  "incidentRecommended": <true|false>,
  "incidentType": "<string>",
  "suggestedAction": "<string>"
}\nThe person count detected by Vertex AI Vision is ${personCount}.\nIf peopleCount is above 75% of capacity, set potentialRisk to true.\nIf peopleCount exceeds capacity, or if smoke/fire/medicalEmergency is detected, set incidentRecommended to true and specify the incidentType.`;

      logger.log("📨 Gemini Prompt:", prompt.substring(0, 200));

      // Fetch zone data
      const zoneSnap = await db.collection("zones").doc(zoneId).get();
      if (!zoneSnap.exists) {
        logger.log(`⚠️ No zone doc found for ${zoneId}`);
        return;
      }
      const zoneData = zoneSnap.data();
      const zoneCapacity = zoneData.capacity || 100;
      const zoneLocation = zoneData.boundaries?.coordinates?.[0] || {};
      const zoneName = zoneData.name || zoneId;

      // Call Gemini
      const response = await geminiModel.generateContent({
        contents: [
          {
            role: "user",
            parts: [
              { text: prompt },
              {
                inlineData: {
                  mimeType,
                  data: base64Image,
                },
              },
            ],
          },
        ],
        generationConfig: {
          maxOutputTokens: null,
          temperature: 0.2,
          responseMimeType: "application/json",
        },
      });

      logger.log("📨 Gemini raw response:", JSON.stringify(response));

      const text =
        response?.response?.candidates?.[0]?.content?.parts?.[0]?.text || "";

      let crowdDensity = "low";
      let smokeDetected = false;
      let fireDetected = false;
      let medicalEmergency = false;
      let potentialRisk = false;
      let incidentRecommended = false;
      let incidentType = null;
      let suggestedAction = "No action needed";
      try {
        const json = JSON.parse(text);
        crowdDensity = json.crowdDensity ?? "low";
        smokeDetected = json.smokeDetected ?? false;
        fireDetected = json.fireDetected ?? false;
        medicalEmergency = json.medicalEmergency ?? false;
        potentialRisk = json.potentialRisk ?? false;
        incidentRecommended = json.incidentRecommended ?? false;
        incidentType = json.incidentType ?? null;
        suggestedAction = json.suggestedAction ?? suggestedAction;
      } catch (err) {
        logger.error("⚠️ Failed to parse Gemini response JSON", err.message);
        return;
      }

      // Always update zone with latest crowd data and analysis
      await db.collection('zones').doc(zoneId).update({
        lastPersonCount: personCount,
        lastCrowdDensity: crowdDensity,
        lastAnalyzedAt: admin.firestore.FieldValue.serverTimestamp(),
        lastSmokeDetected: smokeDetected,
        lastFireDetected: fireDetected,
        lastMedicalEmergency: medicalEmergency,
        lastPotentialRisk: potentialRisk,
        lastIncidentRecommended: incidentRecommended,
        lastIncidentType: incidentType,
        lastSuggestedAction: suggestedAction,
        lastImageUrl: publicImageUrl,
      });
      // Add to crowdHistory subcollection
      await db.collection('zones').doc(zoneId).collection('crowdHistory').add({
        peopleCount: personCount,
        crowdDensity,
        timestamp: admin.firestore.FieldValue.serverTimestamp(),
        source: "Vertex Vision",
        imageUrl: publicImageUrl,
      });
      // Decision logic: create incident if Gemini recommends
      if (incidentRecommended && incidentType) {
        let incidentDesc = [];
        let tags = [];
        let priority = "medium";
        let severity = 3;
        if (incidentType === "crowd") {
          if (personCount > zoneCapacity) {
            incidentDesc.push(`Overcrowding detected: ${personCount} > capacity ${zoneCapacity}`);
            tags.push("overcrowding");
            priority = "high";
            severity = 4;
          } else if (potentialRisk) {
            incidentDesc.push(`Potential crowd risk: ${personCount} / ${zoneCapacity}`);
            tags.push("potential-crowd-risk");
            priority = "medium";
            severity = 3;
          }
        }
        if (incidentType === "fire") {
          if (fireDetected) {
            incidentDesc.push("Fire detected");
            tags.push("fire");
            priority = "critical";
            severity = 5;
          } else if (smokeDetected) {
            incidentDesc.push("Smoke detected");
            tags.push("smoke");
            priority = "critical";
            severity = 5;
          }
        }
        if (incidentType === "medical") {
          if (medicalEmergency) {
            incidentDesc.push("Medical emergency detected");
            tags.push("medical");
            priority = "high";
            severity = 4;
          }
        }

        // Find available responder for the zone/type
        let assignedResponderId = null;
        let assignedResponderDoc = null;
        let responderTypeQuery = (incidentType === "fire") ? "Fire Brigade" : (incidentType === "medical") ? "Medical" : "Security";
        let respondersSnap = await db.collection("responders")
          .where("status", "==", "available")
          .where("type", "==", responderTypeQuery)
          .get();
        let responderDocs = respondersSnap.docs;
        // If no match, fallback to any available responder
        if (responderDocs.length === 0) {
          respondersSnap = await db.collection("responders")
            .where("status", "==", "available")
            .get();
          responderDocs = respondersSnap.docs;
        }
        if (responderDocs.length > 0) {
          const randomIdx = Math.floor(Math.random() * responderDocs.length);
          assignedResponderDoc = responderDocs[randomIdx];
          assignedResponderId = assignedResponderDoc.id;
          // Mark responder as assigned
          await assignedResponderDoc.ref.update({
            status: "en_route",
            assignedIncident: null, // will update after incident creation
          });
        }
        // Create incident
        const incidentRef = await db.collection("incidents").add({
          type: incidentType,
          status: "active",
          priority,
          severity,
          zoneId,
          location: {
            lat: zoneLocation.lat || null,
            lng: zoneLocation.lng || null,
            address: zoneName,
          },
          description: incidentDesc.join("; ") || suggestedAction,
          timestamp: admin.firestore.FieldValue.serverTimestamp(),
          reportedBy: "system-vertex-gemini",
          assignedResponderId: assignedResponderId || null,
          assignmentTransactionId: null, // will update after assignment
          environmentalData: {},
          crowdData: {
            density: crowdDensity,
            peopleCount: personCount,
          },
          equipment: [],
          tags,
          media: {
            photos: [publicImageUrl],
          },
          lastUpdated: admin.firestore.FieldValue.serverTimestamp(),
          resolvedAt: null,
          autoGenerated: true,
          source: "Vertex+Gemini Vision",
        });
        // Update zone with latest crowd data and analysis
        // await db.collection('zones').doc(zoneId).update({
        //   lastPersonCount: personCount,
        //   lastCrowdDensity: crowdDensity,
        //   lastAnalyzedAt: admin.firestore.FieldValue.serverTimestamp(),
        //   lastSmokeDetected: smokeDetected,
        //   lastFireDetected: fireDetected,
        //   lastMedicalEmergency: medicalEmergency,
        //   lastPotentialRisk: potentialRisk,
        //   lastIncidentRecommended: incidentRecommended,
        //   lastIncidentType: incidentType,
        //   lastSuggestedAction: suggestedAction,
        //   lastImageUrl: publicImageUrl,
        // });
        // Create responder assignment if responder assigned
        if (assignedResponderId) {
          const assignmentRef = await db.collection("responderAssignments").add({
            incidentId: incidentRef.id,
            responderId: assignedResponderId,
            assignedAt: admin.firestore.FieldValue.serverTimestamp(),
            statusHistory: [
              { status: "en_route", timestamp: new Date().toISOString() },
            ],
            locationHistory: [
              { lat: zoneLocation.lat || null, lng: zoneLocation.lng || null, timestamp: new Date().toISOString() },
            ],
          });
          // Update incident with assignmentTransactionId
          await incidentRef.update({ assignmentTransactionId: assignmentRef.id });
          // Update responder with assignedIncident
          await assignedResponderDoc.ref.update({ assignedIncident: incidentRef.id });
        }

        logger.log(`🚨 Incident created for ${zoneId} (type: ${incidentType})`);
      }

      logger.log(
        `✅ Vertex+Gemini analysis complete: people=${personCount}, density=${crowdDensity}, smoke=${smokeDetected}, fire=${fireDetected}, medical=${medicalEmergency}`
      );
    } catch (err) {
      logger.error('Vertex+Gemini analysis failed', err);
    }
  }
); 

File: drishti-mvp-adk-service/analyze_with_vertex_vision.py
import os
from fastapi import APIRouter, HTTPException
from typing import Dict, Any
from google.cloud import aiplatform, firestore, storage, vision
from datetime import datetime
import base64
import json
import requests
import vertexai
from vertexai.generative_models import GenerativeModel, Part

router = APIRouter()

PROJECT = os.getenv("GCP_PROJECT")
LOCATION = os.getenv("GCP_REGION", "us-central1")
GEMINI_MODEL = os.getenv("GEMINI_MODEL", "gemini-2.5-pro")

firestore_db = firestore.Client()
storage_client = storage.Client()
vision_client = vision.ImageAnnotatorClient()

def download_image_as_base64(bucket_name, object_path):
    bucket = storage_client.bucket(bucket_name)
    blob = bucket.blob(object_path)
    image_bytes = blob.download_as_bytes()
    return base64.b64encode(image_bytes).decode("utf-8"), image_bytes

@router.post("/analyze/vertex-vision")
def analyze_with_vertex_vision(request: Dict):
    # Accept either image (base64), gcs_path, or image_url
    image_b64 = request.get("image")
    gcs_path = request.get("gcs_path")
    image_url = request.get("image_url")
    media_id = request.get("media_id")
    prompt_text = request.get("text")

    # 1. If GCS path is provided, download and base64 encode
    if gcs_path:
        if gcs_path.startswith("gs://"):
            gcs_path = gcs_path[5:]
        bucket_name, object_path = gcs_path.split("/", 1)
        image_b64, image_bytes = download_image_as_base64(bucket_name, object_path)
    elif image_url:
        resp = requests.get(image_url)
        if resp.status_code != 200:
            raise HTTPException(status_code=400, detail="Failed to download image from URL.")
        image_bytes = resp.content
        image_b64 = base64.b64encode(image_bytes).decode("utf-8")
    elif image_b64:
        image_bytes = base64.b64decode(image_b64)
    else:
        raise HTTPException(status_code=400, detail="Provide at least image, gcs_path, or image_url.")

    # 2. Fetch Firestore metadata if media_id or gcs_path is provided
    media_doc = None
    zone_doc = None
    if media_id:
        media_ref = firestore_db.collection("media").document(media_id)
        media_doc = media_ref.get()
        if media_doc.exists:
            media_data = media_doc.to_dict()
            zone_id = media_data.get("zone")
            if zone_id:
                zone_ref = firestore_db.collection("zones").document(zone_id)
                zone_doc = zone_ref.get()
    elif gcs_path:
        media_query = firestore_db.collection("media").where("objectPath", "==", object_path).limit(1).stream()
        for doc in media_query:
            media_doc = doc
            media_data = doc.to_dict()
            zone_id = media_data.get("zone")
            if zone_id:
                zone_ref = firestore_db.collection("zones").document(zone_id)
                zone_doc = zone_ref.get()
            break

    # 3. Call Vertex Vision for object localization (person count)
    person_count = 0
    try:
        vision_response = vision_client.object_localization({"image": {"content": image_bytes}})
        objects = vision_response.localized_object_annotations
        person_count = sum(1 for obj in objects if obj.name.lower() == "person")
    except Exception as e:
        print("Vertex Vision error:", e)

    # 4. Build structured prompt for Gemini
    zone_context = ""
    if zone_doc and zone_doc.exists:
        zone_data = zone_doc.to_dict()
        zone_context = f"Zone: {json.dumps(zone_data)}"
    else:
        zone_context = "Zone: Unknown"

    schema = '''Analyze the image for the following and return ONLY this JSON:\n{\n  "peopleCount": <number>,\n  "crowdDensity": "low|moderate|high",\n  "smokeDetected": <true|false>,\n  "fireDetected": <true|false>,\n  "medicalEmergency": <true|false>,\n  "potentialRisk": <true|false>,\n  "incidentRecommended": <true|false>,\n  "incidentType": "<string>",\n  "suggestedAction": "<string>"\n}\n'''

    prompt = f"""
{prompt_text or ''}
{zone_context}
The person count detected by Vertex AI Vision is {person_count}.
{schema}
"""

    # 5. Call Gemini
    vertexai.init(project=PROJECT, location=LOCATION)
    model = GenerativeModel(GEMINI_MODEL)
    contents = [
        Part.from_text(prompt),
        Part.from_image(image_bytes) if image_bytes else Part.from_text("")
    ]
    response = model.generate_content(contents)
    # 6. Parse JSON from Gemini response
    try:
        text = response.text
        text = text.replace("```json", "").replace("```", "").strip()
        result_json = json.loads(text)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Gemini response parse error: {e}, raw: {text}")

    # 7. Store result in Firestore (optional, if media_id or gcs_path)
    if media_doc:
        media_ref = firestore_db.collection("media").document(media_doc.id)
        media_ref.update({"vertexVisionAnalysis": result_json, "visionPersonCount": person_count, "analyzedAt": datetime.utcnow()})

    return {"result": result_json, "personCount": person_count, "raw": text} 